# combine the dataframes
In this notebook we gonna combine the dataframes into one common dataframe, that lists the clusters from the used cluster approaches
The aim is to inspect the clustered data

In [11]:
import pickle
import json
import pandas as pd
import numpy as np


In [18]:
fieldsdf = pd.DataFrame()

with open('obj/onehotdf0518.pkl', 'rb') as handle:
    fieldsdf = pickle.load(handle)

fieldsdf.head()



swagger  info  host  basePath  schemes  \
1forge.com                           1     1     1         1        1   
6-dot-authentiqio.appspot.com        1     1     1         1        1   
adafruit.com                         1     1     1         1        1   
adobe.com:aem                        1     1     1         1        1   
adyen.com:CheckoutService            0     1     0         0        0   

                               consumes  produces  paths  definitions  \
1forge.com                            0         1      1            0   
6-dot-authentiqio.appspot.com         0         0      1            1   
adafruit.com                          0         1      1            1   
adobe.com:aem                         0         0      1            1   
adyen.com:CheckoutService             0         0      1            0   

                               parameters  responses  securityDefinitions  \
1forge.com                              0          0                    0   
6-dot-authentiqio.appspot.com           1          1                    0   
adafruit.com                            1          0                    1   
adobe.com:aem                           0          0                    1   
adyen.com:CheckoutService               0          0                    0   

                               security  tags  externalDocs  
1forge.com                            0     0             0  
6-dot-authentiqio.appspot.com         0     0             0  
adafruit.com                          1     0             0  
adobe.com:aem                         1     0             0  
adyen.com:CheckoutService             0     0             0

In [25]:
infodf = pd.DataFrame()

with open('obj/infohotdf0518.pkl', 'rb') as handle:
    infodf = pickle.load(handle)

infodf.head()


title  description  termsOfService  contact  \
1forge.com                         1            1               0        1   
6-dot-authentiqio.appspot.com      1            1               1        1   
adafruit.com                       1            1               0        0   
adobe.com:aem                      1            1               0        1   
adyen.com:CheckoutService          1            1               0        1   

                               license  version  
1forge.com                           0        1  
6-dot-authentiqio.appspot.com        1        1  
adafruit.com                         0        1  
adobe.com:aem                        0        1  
adyen.com:CheckoutService            0        1

In [26]:
keys_field = list(fieldsdf.keys())
keys_field

['swagger',
 'info',
 'host',
 'basePath',
 'schemes',
 'consumes',
 'produces',
 'paths',
 'definitions',
 'parameters',
 'responses',
 'securityDefinitions',
 'security',
 'tags',
 'externalDocs']

In [27]:
keys_info = list(infodf.keys())
keys_info

['title', 'description', 'termsOfService', 'contact', 'license', 'version']

In [39]:
type(infodf)

pandas.core.frame.DataFrame

In [40]:
type(fieldsdf)

pandas.core.frame.DataFrame

In [41]:
comp_df = pd.concat([fieldsdf, infodf], axis=1)

In [42]:
comp_df.head()

swagger  info  host  basePath  schemes  \
1forge.com                           1     1     1         1        1   
6-dot-authentiqio.appspot.com        1     1     1         1        1   
adafruit.com                         1     1     1         1        1   
adobe.com:aem                        1     1     1         1        1   
adyen.com:CheckoutService            0     1     0         0        0   

                               consumes  produces  paths  definitions  \
1forge.com                            0         1      1            0   
6-dot-authentiqio.appspot.com         0         0      1            1   
adafruit.com                          0         1      1            1   
adobe.com:aem                         0         0      1            1   
adyen.com:CheckoutService             0         0      1            0   

                               parameters   ...     securityDefinitions  \
1forge.com                              0   ...                       0   
6-dot-authentiqio.appspot.com           1   ...                       0   
adafruit.com                            1   ...                       1   
adobe.com:aem                           0   ...                       1   
adyen.com:CheckoutService               0   ...                       0   

                               security  tags  externalDocs  title  \
1forge.com                            0     0             0      1   
6-dot-authentiqio.appspot.com         0     0             0      1   
adafruit.com                          1     0             0      1   
adobe.com:aem                         1     0             0      1   
adyen.com:CheckoutService             0     0             0      1   

                               description  termsOfService  contact  license  \
1forge.com                               1               0        1        0   
6-dot-authentiqio.appspot.com            1               1        1        1   
adafruit.com                             1               0        0        0   
adobe.com:aem                            1               0        1        0   
adyen.com:CheckoutService                1               0        1        0   

                               version  
1forge.com                           1  
6-dot-authentiqio.appspot.com        1  
adafruit.com                         1  
adobe.com:aem                        1  
adyen.com:CheckoutService            1  

[5 rows x 21 columns]

In [43]:
comp_df.to_csv('comp_df0518.csv', encoding='utf8')

In [44]:
with open('obj/'+ 'comp_df0518' + '.pkl', 'wb') as f:
        pickle.dump(comp_df, f, pickle.HIGHEST_PROTOCOL)


# dataframe cluster structure complete

In [18]:
pamcompdf = pd.read_csv('PAMcomplete0518.csv',index_col='X')
pamcompdf = pamcompdf.iloc[:,1:]
pamcompdf.head()

swagger  info  host  basePath  schemes  \
X                                                                       
1forge.com                           1     1     1         1        1   
6-dot-authentiqio.appspot.com        1     1     1         1        1   
adafruit.com                         1     1     1         1        1   
adobe.com:aem                        1     1     1         1        1   
afterbanks.com                       1     1     1         1        1   

                               consumes  produces  paths  definitions  \
X                                                                       
1forge.com                            0         1      1            0   
6-dot-authentiqio.appspot.com         0         0      1            1   
adafruit.com                          0         1      1            1   
adobe.com:aem                         0         0      1            1   
afterbanks.com                        0         1      1            1   

                               parameters         ...          security  tags  \
X                                                 ...                           
1forge.com                              0         ...                 0     0   
6-dot-authentiqio.appspot.com           1         ...                 0     0   
adafruit.com                            1         ...                 1     0   
adobe.com:aem                           0         ...                 1     0   
afterbanks.com                          0         ...                 0     0   

                               externalDocs  title  description  \
X                                                                 
1forge.com                                0      1            1   
6-dot-authentiqio.appspot.com             0      1            1   
adafruit.com                              0      1            1   
adobe.com:aem                             0      1            1   
afterbanks.com                            0      1            1   

                               termsOfService  contact  license  version  \
X                                                                          
1forge.com                                  0        1        0        1   
6-dot-authentiqio.appspot.com               1        1        1        1   
adafruit.com                                0        0        0        1   
adobe.com:aem                               0        1        0        1   
afterbanks.com                              0        0        0        1   

                               pam_fit$clustering  
X                                                  
1forge.com                                      1  
6-dot-authentiqio.appspot.com                   1  
adafruit.com                                    2  
adobe.com:aem                                   2  
afterbanks.com                                  1  

[5 rows x 22 columns]

  X swagger info host basePath schemes consumes produces paths definitions parameters responses securityDefinitions security
  tags externalDocs title description termsOfService contact license version
- 823        landregistry.gov.uk:deed       1    1    1        1       1        0        1     1           1          0         0                   0        0     0            0     1           1              0       1       0       1
- 980                   voodoomfg.com       1    1    1        1       1        0        1     1           1          0         0                   1        1 0            0     1           1              0       1       0       1
- 861                    neowsapp.com       1    1    1        1       1        1        1     1           1          0         0                   0        0 1            0     1           1              0       0       0       1
- 140              amazonaws.com:xray       1    1    1        1       1        1        1     1           1          1         0                   1        1 0            1     1           1              1       1       1       1
- 777 googleapis.com:youtubereporting       1    1    1        1       1        0        0     1           1          1         0                   1        0 1            1     1           1              0       1       0       1
- 997           windows.net:graphrbac       1    1    1        0       1        1        1     1           1          1         0                   1        1  0            0     1           1              0       0       0       1
    
  
 

<img src="clusterF1.png">

In [19]:
pamcompdf.loc[pamcompdf['pam_fit$clustering'] == 6]

swagger  info  host  \
X                                                                         
azure.com:advisor                                         1     1     1   
azure.com:analysisservices                                1     1     1   
azure.com:apimanagement                                   1     1     1   
azure.com:apimanagement-apimapis                          1     1     1   
azure.com:apimanagement-apimauthorizationservers          1     1     1   
azure.com:apimanagement-apimbackends                      1     1     1   
azure.com:apimanagement-apimcertificates                  1     1     1   
azure.com:apimanagement-apimdeployment                    1     1     1   
azure.com:apimanagement-apimdiagnostics                   1     1     1   
azure.com:apimanagement-apimemailtemplate                 1     1     1   
azure.com:apimanagement-apimgroups                        1     1     1   
azure.com:apimanagement-apimidentityprovider              1     1     1   
azure.com:apimanagement-apimloggers                       1     1     1   
azure.com:apimanagement-apimnetworkstatus                 1     1     1   
azure.com:apimanagement-apimnotifications                 1     1     1   
azure.com:apimanagement-apimopenidconnectproviders        1     1     1   
azure.com:apimanagement-apimproducts                      1     1     1   
azure.com:apimanagement-apimproperties                    1     1     1   
azure.com:apimanagement-apimquotas                        1     1     1   
azure.com:apimanagement-apimreports                       1     1     1   
azure.com:apimanagement-apimsubscriptions                 1     1     1   
azure.com:apimanagement-apimtagresources                  1     1     1   
azure.com:apimanagement-apimtags                          1     1     1   
azure.com:apimanagement-apimtenant                        1     1     1   
azure.com:apimanagement-apimusers                         1     1     1   
azure.com:apimanagement-apimversionsets                   1     1     1   
azure.com:applicationinsights-aiOperations_API            1     1     1   
azure.com:applicationinsights-componentAnnotati...        1     1     1   
azure.com:applicationinsights-componentApiKeys_API        1     1     1   
azure.com:applicationinsights-componentContinuo...        1     1     1   
...                                                     ...   ...   ...   
azure.com:streamanalytics-subscriptions                   1     1     1   
azure.com:streamanalytics-transformations                 1     1     1   
azure.com:subscription-subscriptionDefinitions            1     1     1   
azure.com:timeseriesinsights                              1     1     1   
azure.com:trafficmanager                                  1     1     1   
azure.com:trafficmanager-trafficmanageranalytics          1     1     1   
azure.com:visualstudio-Csm                                1     1     1   
azure.com:web-AppServiceCertificateOrders                 1     1     1   
azure.com:web-AppServiceEnvironments                      1     1     1   
azure.com:web-AppServicePlans                             1     1     1   
azure.com:web-CertificateRegistrationProvider             1     1     1   
azure.com:web-Certificates                                1     1     1   
azure.com:web-DeletedWebApps                              1     1     1   
azure.com:web-Diagnostics                                 1     1     1   
azure.com:web-DomainRegistrationProvider                  1     1     1   
azure.com:web-Domains                                     1     1     1   
azure.com:web-Provider                                    1     1     1   
azure.com:web-Recommendations                             1     1     1   
azure.com:web-ResourceProvider                            1     1     1   
azure.com:web-TopLevelDomains                             1     1     1   
azure.com:web-WebApps                                     1     1     1   
az

In [20]:
for i in range(1,7):
    print("#########")
    print("cluster" + str(i))
    print("#########")
    print(str(pamcompdf.loc[pamcompdf['pam_fit$clustering'] == i].count()))


#########
cluster1
#########
swagger                139
info                   139
host                   139
basePath               139
schemes                139
consumes               139
produces               139
paths                  139
definitions            139
parameters             139
responses              139
securityDefinitions    139
security               139
tags                   139
externalDocs           139
title                  139
description            139
termsOfService         139
contact                139
license                139
version                139
pam_fit$clustering     139
dtype: int64
#########
cluster2
#########
swagger                113
info                   113
host                   113
basePath               113
schemes                113
consumes               113
produces               113
paths                  113
definitions            113
parameters             113
responses              113
securityDefinitions    113
security   

# dataframe clusters strucutre info

In [21]:
infodf = pd.read_csv('PAMinfo.csv',index_col='X')
infodf = infodf.iloc[:,1:]
infodf.head()

title  description  termsOfService  contact  \
X                                                                            
1forge.com                         1            1               0        1   
6-dot-authentiqio.appspot.com      1            1               1        1   
adafruit.com                       1            1               0        0   
adobe.com:aem                      1            1               0        1   
afterbanks.com                     1            1               0        0   

                               license  version  used.keys  missing.keys  \
X                                                                          
1forge.com                           0        1          4             2   
6-dot-authentiqio.appspot.com        1        1          6             0   
adafruit.com                         0        1          3             3   
adobe.com:aem                        0        1          4             2   
afterbanks.com                       0        1          3             3   

                               pam_fit$clustering  
X                                                  
1forge.com                                      1  
6-dot-authentiqio.appspot.com                   2  
adafruit.com                                    3  
adobe.com:aem                                   1  
afterbanks.com                                  3

clusterdata[pam_fit$medoids, ]
                       
1016            zuora.com

1007            wowza.com

1015      zoomconnect.com

1014              zoom.us

1005 wordassociations.net

<img src="clusterF2.png">

In [22]:
infodf.loc[infodf['pam_fit$clustering'] == 1].count()

title                 240
description           240
termsOfService        240
contact               240
license               240
version               240
used.keys             240
missing.keys          240
pam_fit$clustering    240
dtype: int64

In [23]:
infodf.loc[infodf['pam_fit$clustering'] == 3]

title  description  termsOfService  \
X                                                                              
adafruit.com                                  1            1               0   
afterbanks.com                                1            1               0   
airport-web.appspot.com                       1            1               0   
apacta.com                                    1            1               0   
apigee.net:marketcheck-cars                   1            1               0   
apitore.com:2chMatomeFeedsApis                1            1               0   
apitore.com:certificateFeedsApis              1            1               0   
apitore.com:cosineSimilarity                  1            1               0   
apitore.com:designFeedsApis                   1            1               0   
apitore.com:documentFrequencyApis             1            1               0   
apitore.com:gameFeedsApis                     1            1               0   
apitore.com:japaneseWordnetApis               1            1               0   
apitore.com:kmeansClusteringByWord2vec        1            1               0   
apitore.com:kuromojiApis                      1            1               0   
apitore.com:languageDetectionApis             1            1               0   
apitore.com:newsFeedsApis                     1            1               0   
apitore.com:pioneerBlogFeedsApis              1            1               0   
apitore.com:rss2jsonApis                      1            1               0   
apitore.com:sentenceSeparatorApis             1            1               0   
apitore.com:sentenceSimilarity                1            1               0   
apitore.com:sentencepieceApis_tweet_          1            1               0   
apitore.com:sentencepieceApis_wikipedia_      1            1               0   
apitore.com:sentimentApis                     1            1               0   
apitore.com:synonymApis                       1            1               0   
apitore.com:techFeedsApis                     1            1               0   
apitore.com:text2labelByTfidfApis             1            1               0   
apitore.com:text2labelByWordVectorApis        1            1               0   
apitore.com:travelFeedsApis                   1            1               0   
apitore.com:tweetSummarizationApis            1            1               0   
apitore.com:twitterApis                       1            1               0   
...                                         ...          ...             ...   
smart-me.com                                  1            1               0   
spotify.com                                   1            0               0   
stackexchange.com                             1            1               0   
staging-ecotaco.com                           1            1               0   
statsocial.com                                1            1               0   
synq.fm                                       1            1               0   
taxamo.com                                    1            1               0   
tfl.gov.uk                                    1            1               0   
thenounproject.com                            1            1               0   
thetvdb.com                                   1            1               0   
traccar.org                                   1            1               0   
twilio.com                                    1            1               0   
vatapi.com                                    1            1               0   
versioneye.com                                1            1               0   
waag.org                                      1            1               0   
watchful.li                                   1            0               0   
weber-gesamtausgabe.de                        1            1               0   
windows.net:batch-BatchService            

# dataframe clusters path metrics

In [24]:
pathdf = pd.read_csv('PAMpath.csv',index_col='X')
pathdf = pathdf.iloc[:,1:]

pathdf.head()

paths  operations  pam_fit$clustering
X                                                                   
1forge.com                         2           2                   1
6-dot-authentiqio.appspot.com      5          14                   2
adafruit.com                      34          69                   3
adobe.com:aem                     26          30                   4
afterbanks.com                     3           3                   5

                       X paths operations
- 1010                                             xkcd.com     2          2
- 239         azure.com:authorization-authorization-RACalls     6         10
- 922                                               ritc.io    36         66
- 976                                          vestorly.com    21         33
- 992                                    whapi.com:sessions     3          4
- 528                                       box.com:content    77        127
- 994                                    who-hosts-this.com     1          1
- 240  azure.com:authorization-authorization-RoleBasedCalls    12         18
- 196                                    aucklandmuseum.com     4          6
- 511                                 azure.com:web-WebApps   224        311

<img src="clusterF3.png">

In [25]:
pathdf.loc[pathdf['pam_fit$clustering'] == 2]

paths  operations  \
X                                                                       
6-dot-authentiqio.appspot.com                           5          14   
amazonaws.com:autoscaling-plans                         4           8   
amazonaws.com:ce                                        5          10   
amazonaws.com:health                                    6          12   
amazonaws.com:importexport                              6          12   
amazonaws.com:polly                                     4          10   
amazonaws.com:resourcegroupstaggingapi                  5          10   
amazonaws.com:streams.dynamodb                          4           8   
amazonaws.com:xray                                      6          12   
apache.org:qakka                                        7          10   
apitore.com:wordnetApis                                 8           8   
apitore.com:wordnetSimilarityApis                       8           8   
applicationinsights.io:applicationinsights-AppI...      8           9   
authentiq.io                                            6           9   
azure.com:advisor                                       9          13   
azure.com:apimanagement-apimdiagnostics                 4          10   
azure.com:apimanagement-apimgroups                      4          10   
azure.com:apimanagement-apimnotifications               6          11   
azure.com:apimanagement-apimreports                     8           8   
azure.com:apimanagement-apimsubscriptions               4           8   
azure.com:apimanagement-apimtenant                     10          11   
azure.com:apimanagement-apimusers                       7          11   
azure.com:authorization-authorization-RACalls           6          10   
azure.com:automation-account                            6           9   
azure.com:automation-dscNode                            7           9   
azure.com:automation-job                                9          10   
azure.com:automation-module                             7          10   
azure.com:cognitiveservices                             8          11   
azure.com:cognitiveservices-ComputerVision              9           9   
azure.com:datalake-analytics-job                       11          13   
...                                                   ...         ...   
googleapis.com:testing                                  5          10   
googleapis.com:tracing                                  4           8   
googleapis.com:translate                                3           8   
googleapis.com:youtubeAnalytics                         3          11   
googleapis.com:youtubereporting                         6          14   
highwaysengland.co.uk                                  10          10   
inboxroute.com                                          5           8   
instawell.com                                           8           8   
isendpro.com                                           10          11   
magick.nu                                               8           8   
nytimes.com:books_api                                   6          12   
nytimes.com:community                                   4           8   
oceandrivers.com                                        8           8   
openbanking.org.uk                                      6          12   
openstf.io                                              8          10   
owler.com                                              12          12   
pandorabots.com                                         7          13   
payments.service.gov.uk:payments                        6           8   
polygon.io                                             10          10   
qualpay.com                                            10          10   
rbaskets.in                                             4           9   
roaring.io                                              7          11   
slideroom.com                        

In [30]:
for i in range(1,11):
    print("#########")
    print("cluster" + str(i))
    print("#########")
    print(str(pathdf.loc[pathdf['pam_fit$clustering'] == i].count()))



#########
cluster1
#########
paths                 91
operations            91
pam_fit$clustering    91
dtype: int64
#########
cluster2
#########
paths                 128
operations            128
pam_fit$clustering    128
dtype: int64
#########
cluster3
#########
paths                 101
operations            101
pam_fit$clustering    101
dtype: int64
#########
cluster4
#########
paths                 139
operations            139
pam_fit$clustering    139
dtype: int64
#########
cluster5
#########
paths                 107
operations            107
pam_fit$clustering    107
dtype: int64
#########
cluster6
#########
paths                 63
operations            63
pam_fit$clustering    63
dtype: int64
#########
cluster7
#########
paths                 102
operations            102
pam_fit$clustering    102
dtype: int64
#########
cluster8
#########
paths                 148
operations            148
pam_fit$clustering    148
dtype: int64
#########
cluster9
#########
paths            

In [29]:
for i in range (1,11):
    print("#####################")
    print("cluster: "+ str(i))
    print("#####################")
    print("min:" + str(pathdf.loc[pathdf['pam_fit$clustering'] == i].min()))
    print("max:" + str(pathdf.loc[pathdf['pam_fit$clustering'] == i].max()))
   

#####################
cluster: 1
#####################
min:paths                 1
operations            2
pam_fit$clustering    1
dtype: int64
max:paths                 2
operations            2
pam_fit$clustering    1
dtype: int64
#####################
cluster: 2
#####################
min:paths                 3
operations            8
pam_fit$clustering    2
dtype: int64
max:paths                 12
operations            14
pam_fit$clustering     2
dtype: int64
#####################
cluster: 3
#####################
min:paths                 22
operations            42
pam_fit$clustering     3
dtype: int64
max:paths                  68
operations            112
pam_fit$clustering      3
dtype: int64
#####################
cluster: 4
#####################
min:paths                  9
operations            23
pam_fit$clustering     4
dtype: int64
max:paths                 36
operations            51
pam_fit$clustering     4
dtype: int64
#####################
cluster: 5
#################

# dataframe clusters fields  count structure

In [46]:

fieldcompdf = pd.read_csv('PAMcomp.csv',index_col='X')
fieldcompdf = fieldcompdf.iloc[:,1:]
fieldcompdf.head()


swagger  info  host  basePath  schemes  \
X                                                                       
1forge.com                           1     1     1         1        1   
6-dot-authentiqio.appspot.com        1     1     1         1        1   
adafruit.com                         1     1     1         1        1   
adobe.com:aem                        1     1     1         1        1   
afterbanks.com                       1     1     1         1        1   

                               consumes  produces  paths  definitions  \
X                                                                       
1forge.com                            0         1      1            0   
6-dot-authentiqio.appspot.com         0         0      1            1   
adafruit.com                          0         1      1            1   
adobe.com:aem                         0         0      1            1   
afterbanks.com                        0         1      1            1   

                               parameters  responses  securityDefinitions  \
X                                                                           
1forge.com                              0          0                    0   
6-dot-authentiqio.appspot.com           1          1                    0   
adafruit.com                            1          0                    1   
adobe.com:aem                           0          0                    1   
afterbanks.com                          0          0                    0   

                               security  tags  externalDocs  \
X                                                             
1forge.com                            0     0             0   
6-dot-authentiqio.appspot.com         0     0             0   
adafruit.com                          1     0             0   
adobe.com:aem                         1     0             0   
afterbanks.com                        0     0             0   

                               implemented.fields  missing.fields  cluster  \
X                                                                            
1forge.com                                      7               8        0   
6-dot-authentiqio.appspot.com                   9               6        1   
adafruit.com                                   11               4        2   
adobe.com:aem                                   9               6        2   
afterbanks.com                                  8               7        0   

                               pam_fit$clustering  
X                                                  
1forge.com                                      1  
6-dot-authentiqio.appspot.com                   2  
adafruit.com                                    3  
adobe.com:aem                                   2  
afterbanks.com                                  1

clusterdata[pam_fit$medoids, ]

                      X swagger info host basePath schemes consumes produces paths definitions parameters responses
                      
1011          yunbi.com       1    1    1        1       1        0        1     1           0          0         0

1013        zappiti.com       1    1    1        1       1        1        1     1           1          0         0

1016          zuora.com       1    1    1        1       1        1        1     1           1          1         0

1015    zoomconnect.com       1    1    1        1       1        0        1     1           1          0         0

925  rottentomatoes.com       1    1    1        1       1        1        1     1           1          0         0

<img src="clusterF4.png">

In [73]:
fieldcompdf.loc[fieldcompdf['pam_fit$clustering'] == 4]

swagger  info  host  \
X                                                                         
airport-web.appspot.com                                   1     1     1   
amadeus.com                                               1     1     1   
apigee.net:marketcheck-cars                               1     1     1   
apitore.com:2chMatomeFeedsApis                            1     1     1   
apitore.com:certificateFeedsApis                          1     1     1   
apitore.com:cosineSimilarity                              1     1     1   
apitore.com:designFeedsApis                               1     1     1   
apitore.com:documentFrequencyApis                         1     1     1   
apitore.com:gameFeedsApis                                 1     1     1   
apitore.com:japaneseWordnetApis                           1     1     1   
apitore.com:kmeansClusteringByWord2vec                    1     1     1   
apitore.com:kuromojiApis                                  1     1     1   
apitore.com:languageDetectionApis                         1     1     1   
apitore.com:newsFeedsApis                                 1     1     1   
apitore.com:pioneerBlogFeedsApis                          1     1     1   
apitore.com:rss2jsonApis                                  1     1     1   
apitore.com:sentenceSeparatorApis                         1     1     1   
apitore.com:sentenceSimilarity                            1     1     1   
apitore.com:sentencepieceApis_tweet_                      1     1     1   
apitore.com:sentencepieceApis_wikipedia_                  1     1     1   
apitore.com:sentimentApis                                 1     1     1   
apitore.com:synonymApis                                   1     1     1   
apitore.com:techFeedsApis                                 1     1     1   
apitore.com:text2labelByTfidfApis                         1     1     1   
apitore.com:text2labelByWordVectorApis                    1     1     1   
apitore.com:travelFeedsApis                               1     1     1   
apitore.com:tweetSummarizationApis                        1     1     1   
apitore.com:twitterApis                                   1     1     1   
apitore.com:twitterApisWithSentimentByApitoreSe...        1     1     1   
apitore.com:twitterApisWithWord2vecQueryExpansion         1     1     1   
...                                                     ...   ...   ...   
nrel.gov:building-case-studies                            1     1     1   
nrel.gov:transportation-incentives-laws                   1     1     1   
nytimes.com:times_tags                                    1     1     1   
opencagedata.com                                          1     1     1   
openfintech.io                                            1     1     1   
paylocity.com                                             1     1     1   
payrun.io                                                 1     1     1   
personio.de:personnel                                     1     1     1   
postmarkapp.com:account                                   1     1     1   
postmarkapp.com:server                                    1     1     1   
ritc.io                                                   1     1     1   
semantria.com                                             1     1     1   
sendgrid.com                                              1     1     1   
simplyrets.com                                            1     1     1   
slack.com                                                 1     1     1   
slicebox.local                                            1     1     1   
spectrocoin.com                                           1     1     1   
squareup.com                                              1     1     1   
taxrates.io                                               1     1     1   
thesmsworks.co.uk                                         1     1     1   
transavia.com                                             1     1     1   
tr

In [59]:
fieldstructdf.loc[fieldstructdf['pam_fit$clustering'] == 3].head()

swagger  info  host  basePath  schemes  \
X                                                                               
amazonaws.com:AWSMigrationHub                1     1     1         1        1   
amazonaws.com:acm                            1     1     1         1        1   
amazonaws.com:alexaforbusiness               1     1     1         1        1   
amazonaws.com:apigateway                     1     1     1         1        1   
amazonaws.com:application-autoscaling        1     1     1         1        1   

                                       consumes  produces  paths  definitions  \
X                                                                               
amazonaws.com:AWSMigrationHub                 1         1      1            1   
amazonaws.com:acm                             1         1      1            1   
amazonaws.com:alexaforbusiness                1         1      1            1   
amazonaws.com:apigateway                      1         1      1            1   
amazonaws.com:application-autoscaling         1         1      1            1   

                                       parameters  responses  \
X                                                              
amazonaws.com:AWSMigrationHub                   1          0   
amazonaws.com:acm                               1          0   
amazonaws.com:alexaforbusiness                  1          0   
amazonaws.com:apigateway                        1          0   
amazonaws.com:application-autoscaling           1          0   

                                       securityDefinitions  security  tags  \
X                                                                            
amazonaws.com:AWSMigrationHub                            1         1     0   
amazonaws.com:acm                                        1         1     0   
amazonaws.com:alexaforbusiness                           1         1     0   
amazonaws.com:apigateway                                 1         1     0   
amazonaws.com:application-autoscaling                    1         1     0   

                                       externalDocs  pam_fit$clustering  
X                                                                        
amazonaws.com:AWSMigrationHub                     1                   3  
amazonaws.com:acm                                 1                   3  
amazonaws.com:alexaforbusiness                    1                   3  
amazonaws.com:apigateway                          1                   3  
amazonaws.com:application-autoscaling             1                   3

In [60]:
for i in range(1,6):
    print("#########")
    print("cluster" + str(i))
    print("#########")
    print(str(fieldcompdf.loc[fieldcompdf['pam_fit$clustering'] == i].count()))


#########
cluster1
#########
swagger                99
info                   99
host                   99
basePath               99
schemes                99
consumes               99
produces               99
paths                  99
definitions            99
parameters             99
responses              99
securityDefinitions    99
security               99
tags                   99
externalDocs           99
implemented.fields     99
missing.fields         99
cluster                99
pam_fit$clustering     99
dtype: int64
#########
cluster2
#########
swagger                91
info                   91
host                   91
basePath               91
schemes                91
consumes               91
produces               91
paths                  91
definitions            91
parameters             91
responses              91
securityDefinitions    91
security               91
tags                   91
externalDocs           91
implemented.fields     91
missing.fields     

In [61]:
for i in range (1,6):
    print("#####################")
    print("cluster: "+ str(i))
    print("#####################")
    print("min:" + str(fieldcompdf.loc[fieldcompdf['pam_fit$clustering'] == i].min()))
    print("##")
    print("max:" + str(fieldcompdf.loc[fieldcompdf['pam_fit$clustering'] == i].max()))
   

#####################
cluster: 1
#####################
min:swagger                0
info                   1
host                   0
basePath               0
schemes                0
consumes               0
produces               0
paths                  1
definitions            0
parameters             0
responses              0
securityDefinitions    0
security               0
tags                   0
externalDocs           0
implemented.fields     3
missing.fields         7
cluster                0
pam_fit$clustering     1
dtype: int64
##
max:swagger                 1
info                    1
host                    1
basePath                1
schemes                 1
consumes                1
produces                1
paths                   1
definitions             1
parameters              1
responses               0
securityDefinitions     1
security                1
tags                    1
externalDocs            1
implemented.fields      8
missing.fields         12
clus

# dataframe fields structure clusters

In [54]:
fieldstructdf = pd.read_csv('PAMStruct.csv',index_col='X')
fieldstructdf = fieldstructdf.iloc[:,1:]
fieldstructdf.head()

swagger  info  host  basePath  schemes  \
X                                                                       
1forge.com                           1     1     1         1        1   
6-dot-authentiqio.appspot.com        1     1     1         1        1   
adafruit.com                         1     1     1         1        1   
adobe.com:aem                        1     1     1         1        1   
afterbanks.com                       1     1     1         1        1   

                               consumes  produces  paths  definitions  \
X                                                                       
1forge.com                            0         1      1            0   
6-dot-authentiqio.appspot.com         0         0      1            1   
adafruit.com                          0         1      1            1   
adobe.com:aem                         0         0      1            1   
afterbanks.com                        0         1      1            1   

                               parameters  responses  securityDefinitions  \
X                                                                           
1forge.com                              0          0                    0   
6-dot-authentiqio.appspot.com           1          1                    0   
adafruit.com                            1          0                    1   
adobe.com:aem                           0          0                    1   
afterbanks.com                          0          0                    0   

                               security  tags  externalDocs  \
X                                                             
1forge.com                            0     0             0   
6-dot-authentiqio.appspot.com         0     0             0   
adafruit.com                          1     0             0   
adobe.com:aem                         1     0             0   
afterbanks.com                        0     0             0   

                               pam_fit$clustering  
X                                                  
1forge.com                                      1  
6-dot-authentiqio.appspot.com                   1  
adafruit.com                                    2  
adobe.com:aem                                   1  
afterbanks.com                                  1

clusterdata[pam_fit$medoids, ]

                                  X swagger info host basePath schemes consumes produces paths definitions parameters
                                  
198                       avaza.com       1    1    1        0       1        0        1     1           1          0

997           windows.net:graphrbac       1    1    1        0       1        1        1     1           1          1

516                      bbci.co.uk       1    1    1        1       1        1        1     1           1          1

148  apitore.com:2chMatomeFeedsApis       1    1    1        1       1        1        1     1           1          0

777 googleapis.com:youtubereporting       1    1    1        1       1        0        0     1           1          1

<img src="clusterF5.png">

In [27]:
fieldstructdf.loc[fieldstructdf['pam_fit$clustering'] == 2]

NameError: name 'fieldstructdf' is not defined

In [64]:
for i in range(1,6):
    print("#########")
    print("cluster" + str(i))
    print("#########")
    print(str(fieldstructdf.loc[fieldstructdf['pam_fit$clustering'] == i].count()))


#########
cluster1
#########
swagger                197
info                   197
host                   197
basePath               197
schemes                197
consumes               197
produces               197
paths                  197
definitions            197
parameters             197
responses              197
securityDefinitions    197
security               197
tags                   197
externalDocs           197
pam_fit$clustering     197
dtype: int64
#########
cluster2
#########
swagger                338
info                   338
host                   338
basePath               338
schemes                338
consumes               338
produces               338
paths                  338
definitions            338
parameters             338
responses              338
securityDefinitions    338
security               338
tags                   338
externalDocs           338
pam_fit$clustering     338
dtype: int64
#########
cluster3
#########
swagger                

In [63]:
for i in range (1,6):
    print("#####################")
    print("cluster: "+ str(i))
    print("#####################")
    print("min:" + str(fieldstructdf.loc[fieldstructdf['pam_fit$clustering'] == i].min()))
    print("##")
    print("max:" + str(fieldstructdf.loc[fieldstructdf['pam_fit$clustering'] == i].max()))
   

#####################
cluster: 1
#####################
min:swagger                1
info                   1
host                   1
basePath               0
schemes                1
consumes               0
produces               0
paths                  1
definitions            0
parameters             0
responses              0
securityDefinitions    0
security               0
tags                   0
externalDocs           0
pam_fit$clustering     1
dtype: int64
##
max:swagger                1
info                   1
host                   1
basePath               1
schemes                1
consumes               1
produces               1
paths                  1
definitions            1
parameters             1
responses              1
securityDefinitions    1
security               1
tags                   1
externalDocs           1
pam_fit$clustering     1
dtype: int64
#####################
cluster: 2
#####################
min:swagger                1
info                   1